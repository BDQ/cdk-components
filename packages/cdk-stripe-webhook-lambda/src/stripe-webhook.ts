import * as Stripe from 'stripe';

import {
  camelizeKeys,
  customResourceHelper,
  OnCreateHandler,
  OnUpdateHandler,
  OnDeleteHandler,
  ResourceHandler,
  ResourceHandlerReturn,
} from 'custom-resource-helper';

export interface WebhookProps {
  secretKey: string;
  url: string;
  events: string[];
}

const handleCreate: OnCreateHandler = async (
  event,
  _,
): Promise<ResourceHandlerReturn> => {
  const { secretKey, url, events } = camelizeKeys(event.ResourceProperties);

  const stripe = new Stripe(secretKey);

  const responseData = await stripe.webhookEndpoints.create({
    url,
    enabled_events: events,
  });

  const physicalResourceId = responseData.id;

  return {
    physicalResourceId,
    responseData,
  };
};

const handleUpdate: OnUpdateHandler = async (
  event,
  _,
): Promise<ResourceHandlerReturn> => {
  const { secretKey, url, events } = camelizeKeys(event.ResourceProperties);

  const webhookId = event.PhysicalResourceId;

  const stripe = new Stripe(secretKey);

  const responseData = await stripe.webhookEndpoints.update(webhookId, {
    url,
    enabled_events: events,
  });

  const physicalResourceId = responseData.id;

  return {
    physicalResourceId,
    responseData,
  };
};

const handleDelete: OnDeleteHandler = async (event, _): Promise<void> => {
  const { secretKey } = camelizeKeys(event.ResourceProperties);

  const webhookId = event.PhysicalResourceId;

  const stripe = new Stripe(secretKey);

  await stripe.webhookEndpoints.del(webhookId);
};

export const handler = customResourceHelper(
  (): ResourceHandler => ({
    onCreate: handleCreate,
    onUpdate: handleUpdate,
    onDelete: handleDelete,
  }),
);
