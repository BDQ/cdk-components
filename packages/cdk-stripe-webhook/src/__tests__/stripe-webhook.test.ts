import { Stack } from '@aws-cdk/core';
import 'jest-cdk-snapshot';

import { StripeWebhook } from '../stripe-webhook';

describe('cdk-stripe-webhook', (): void => {
    it('snapshot', (): void => {
        const stack = new Stack();

        new StripeWebhook(stack, 'StripeWebhook', {
            secretKey: 'sk_test_1213fdsdfsgfsdgdfgfdg',
            url: 'https://www.example.com/test',
            events: ['charge.failed', 'charge.succeeded'],
        });

        expect(stack).toMatchCdkSnapshot();
    });
});
