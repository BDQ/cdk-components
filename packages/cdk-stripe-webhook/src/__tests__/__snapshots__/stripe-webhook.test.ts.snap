// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-stripe-webhook snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035ArtifactHash29E11907": Object {
      "Description": "Artifact hash for asset \\"628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035\\"",
      "Type": "String",
    },
    "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035S3Bucket5E397CBF": Object {
      "Description": "S3 bucket for asset \\"628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035\\"",
      "Type": "String",
    },
    "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035S3VersionKeyF6CA1E6A": Object {
      "Description": "S3 key for asset version \\"628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomStripeWebhooke9db3870d7934cd296a9efe2e318ebbc3F433A9D": Object {
      "DependsOn": Array [
        "CustomStripeWebhooke9db3870d7934cd296a9efe2e318ebbcServiceRole975E615A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035S3Bucket5E397CBF",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035S3VersionKeyF6CA1E6A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters628d723dd93ac7dd3f8622fb9ebdd881ae92ba6bb6315381f3a9d7f83de52035S3VersionKeyF6CA1E6A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lib/stripe-webhook.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomStripeWebhooke9db3870d7934cd296a9efe2e318ebbcServiceRole975E615A",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomStripeWebhooke9db3870d7934cd296a9efe2e318ebbcServiceRole975E615A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StripeWebhookCustomResource8D1A2FEA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Events": Array [
          "charge.failed",
          "charge.succeeded",
        ],
        "SecretKey": "sk_test_1213fdsdfsgfsdgdfgfdg",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomStripeWebhooke9db3870d7934cd296a9efe2e318ebbc3F433A9D",
            "Arn",
          ],
        },
        "Url": "https://www.example.com/test",
      },
      "Type": "Custom::StripeWebhook",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
